design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,613 @@
+by default during login will be the "most preferred" credential belonging to a valid
+authentication category, at each authentication step.
+
+A note for passwords: Unlike other credentials there should not be more than one password
+that can be configured. Also, among its edition option it should be possible to reset
+(temporarily or not) as is currently the case.
+
+#### Deleting
+
+Deleting credentials can lead to a user being unable to authenticate himself.
+While a user shouldn't be able to put himself in this situation, an admin must be
+able to remove any or all credentials for a user if they have been compromised.
+Considering the problems that may result from incorrectly removing a credential,
+the admin GUI should ask for a validation (e.g. "Are you sure?") before doing
+the delete.
+
+If the administrator and user can remove credentials, the possibility to disable
+credentials may no longer be useful. However, there would also be no problem in keeping it.
+
+#### Creating
+
+For some authentication categories, an admin may be able to create a credential.
+The most obvious of these is the password, as is currently the case. The admin may
+create a new password for a user (temporary or not). If one already exists, it
+overwrites the previous one.
+
+For other authentication categories, it is up to them to determine if a credential
+can be created, and to determine how it can be created.
+
+### Modifications to the REST API
+
+Currently there's no way to get credentials with the REST API. This should change with these
+modifications to reflect the new options for the administrator. The API should function in the
+same manner as the admin console: Credentials can be exposed (except for secret
+values), the metadata edited and the credentials deleted with the same restrictions as described in the previous section.
+
+### Database modifications
+
+The current system has multiple fields to store information about a crendiential. Some
+are common, others are not. For example, a password's secret is the hash of the password,
+an otp's secret is the value shared between Keycloak and the device. However, some information
+is only relevant for a credential of a specific authentication category, for example,
+password credentials contain the salt and the hash iterations, while TOTP may have the
+number of digits and the period. Other authenticators will likely have their own set of
+fixed metadata. This data doesn't need to be queried nor modified, therefore the new
+credential table columns could be:
+```
+-----------------------------
+| ID                        |
+-----------------------------
+| authenticator_category    |
+-----------------------------
+| user_label                |
+-----------------------------
+| secret_data               |
+-----------------------------
+| credential_data           |
+-----------------------------
+```
+
+With `ID` the primary key of the credential, `authenticator_category` a string
```

Comment:
I am not sure about the "secret_data" and "credential_data" and have their format as JSON stored in the DB.

IMO it will be good if we have possibility to introduce custom credential types and if the whole credentials page is "auto-rerendered" in the admin console (and maybe in the account mgmt too). In other words, we can have something like "Credential Storage SPI" and each CredentialStorageProviderFactory can contains list of available configuration options and based on that, the particular credentials page is automatically rendered in the admin console. In other words, the individual options of Credential will be just key-value pairs. I can imagine something similar to what we use for various Component providers (or even for Authenticator Configurations, or UserStorage providers which are also rendered generically). The fact whether credential option is secret or not could be handled with some flag in credential metadata. Just a note that ProviderConfigProperty class already contains the "secret" flag on it and some parts in admin console are rendered based on it. Maybe this can be re-used for the credentials directly.

Reply:
Describing these values as json is actually a result of an earlier discussion with @stianst. Functionally, I don't see any difference between your proposition and using json, aside from the fact that json allows sub-elements. I think that what you propose works both with key-values or json.

Having parts marked secret (however it is done), is necessary I believe. First because it marks which part of the credential should not be exposed or transmitted, but also because it is best (security) practice to have confidential data in a database to be encrypted by an external key to protect against a potential sql injection attacks. I didn't necessarily want to bring this up as part of the design proposal, but no matter which scheme is used, encrypting the secret data should be considered.

Reply:
As you pointed, key-value pairs has avantage for querying. This is probably not an advantage if you always know the user as you can retrieve the credentials of given user and filter them programatically (for example based on credential type and device). However if you don't know the user in advance, it may be beneficial. ATM I don't know about any use-case when you don't know the user, but just wanted to point it in case you can thing about some. For example with X509 or SPNEGO authentication, you don't know the user, but it still works fine as you can retrieve the user from the sent SPNEGO token / X509 certificate.

For the versioning/upgrades and nested attributes, it would work with key-value pairs as well, even if JSON is more friendly for those use-cases.

In shortcut, I don't insist on key-value pairs. JSON would probably works just fine unless we have use-case for a need to query credentials of unknown user as I mentioned above (which we probably don't have).